### PR TITLE
Replace opt-in annotation with extension method

### PR DESCRIPTION
**Pull request by marcphilipp**
## Overview

- **Replace annotation with boolean-returning method**
- **Return enum instead of boolean**
- **Create specialized sub-interface of Extension**

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)

### CODE REVIEW

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+ *
+ * @since 5.12
+ * @see InvocationInterceptor#interceptTestClassConstructor
+ * @see ParameterResolver
+ * @see TestInstancePreConstructCallback
+ * @see TestInstancePostProcessor
+ * @see TestInstanceFactory
+ */
+@API(status = EXPERIMENTAL, since = "5.12")
+public interface TestClassInstanceConstructionParticipatingExtension extends Extension {
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java:31]: :speech_balloon: This enforces that only potentially affected extensions can override the method.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+		/**
+		 * The extension should receive an {@link ExtensionContext} scoped to
+		 * the test class.
+		 *
+		 * @deprecated This behavior will be removed from future versions of
+		 * JUnit and {@link #TEST_SCOPED} will become the default.
+		 */
+		@API(status = DEPRECATED, since = "5.12") //
+		@Deprecated
+		DEFAULT,
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java:112]: :speech_balloon: This is called `DEFAULT` (and not `CLASS_SCOPED` or `LEGACY`) so a configuration parameter can be added in a later PR and to avoid extensions from being able to explicitly opt out from the new behavior.

**Comment by marcphilipp** [comment]
@JojOatXGME I'd appreciate it if you could take a look!

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	 *           Therefore, implementations should be idempotent and avoid side
+	 *           effects. They may, however, cache the result for performance in
+	 *           the {@link ExtensionContext.Store Store} of the supplied
+	 *           {@link ExtensionContext}, if necessary.
+	 * @param rootContext the root extension context to allow inspection of
+	 *                    configuration parameters; never {@code null}
+	 * @since 5.12
+	 */
+	@API(status = EXPERIMENTAL, since = "5.12")
+	default ExtensionContextScope getExtensionContextScopeDuringTestClassInstanceConstruction(
```
Comment (JojOatXGME) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java:87]: :thinking: I don't like the length of the method name, but I also don't have good suggestions.

**Review by JojOatXGME — APPROVED** [review]
Looks good. I still kept a few comments to communicate what I was thinking while reading your PR.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		 * @deprecated This behavior will be removed from future versions of
+		 * JUnit and {@link #TEST_SCOPED} will become the default.
+		 */
+		@API(status = DEPRECATED, since = "5.12") //
+		@Deprecated
+		DEFAULT,
+
+		/**
+		 * The extension should receive an {@link ExtensionContext} scoped to
+		 * the test instance.
```
[the remark this answers is not in the document]
Comment (marcphilipp) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java]: Good point! Revised in f8ca2d3c2046301b48f8642680a69bc3e1294553.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	/**
+	 * {@code ExtensionContextScope} is used to define the scope of the
+	 * {@link ExtensionContext} passed to an extension during the creation of
+	 * test instances.
+	 *
+	 * @since 5.12
+	 * @see TestClassInstanceConstructionParticipatingExtension#getExtensionContextScopeDuringTestClassInstanceConstruction
+	 */
+	@API(status = EXPERIMENTAL, since = "5.12")
+	enum ExtensionContextScope {
```
[the remark this answers is not in the document]
Comment (marcphilipp) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java:101]: Yes, I wanted to make it clear that there's no explicit support for opting out of the new behavior since I have plans to add a configuration parameter that forces all extensions to use the new behavior.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+import static org.apiguardian.api.API.Status.DEPRECATED;
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+
+import org.apiguardian.api.API;
+import org.junit.jupiter.api.TestInstance;
+
+/**
+ * Interface for {@link Extension Extensions} that participate in the
+ * construction of test class instances.
+ *
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java:22]: TODO point out that this interface should not be implemented directly

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java (hunk 1 above)
Comment (marcphilipp) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java:31]: Other naming proposals:
- `TestInstantiationAwareExtension`
- `TestConstructionAwareExtension`

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java (hunk 3 above)
Comment (marcphilipp) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java:87]: Other naming proposals:
- `getTestInstantiationExtensionContextScope`
- `getExtensionContextScopeUsedDuringTestInstantiation`
- ~~`use(s)ExtensionContextScopeDuringTestInstantiation`~~
